docs: mention shebang rewrite in the formula cookbook (#11244)
Homebrew/brew: docs/Formula-Cookbook.md

The symlinks created by [`install_symlink`](https://rubydoc.brew.sh/Pathname#install_symlink-instance_method) are guaranteed to be relative. `ln_s` will only produce a relative symlink when given a relative path.

### Rewriting a script shebang

> In some cases, the formula goal is to install a script written in an interpreted language, such as Python or Perl, and use it as an executable. Homebrew provides utilities to rewrite the shebang of a script so its path matches the interpreter dependency. For example, the [`icdiff` formula](https://github.com/Homebrew/homebrew-core/blob/7beae5ab57c65249403699b2b0700fbccf14e6cb/Formula/icdiff.rb) uses such utility.

Reviewer: ```suggestion
In some cases, the formula goal is to install a script written in an interpreted language, such as Python or Perl, to be used as an executable. Homebrew provides a `rewrite_shebang` method to rewrite the shebang of a script to point to the correct interpreter path. For example, the [`icdiff` formula](https://github.com/Homebrew/homebrew-core/blob/7beae5ab57c65249403699b2b0700fbccf14e6cb/Formula/icdiff.rb#L16) uses such utility.
```

Here's a suggestion. I don't feel that strongly about it.

> In some cases, the formula goal is to install a script written in an interpreted language, such as Python or Perl, to be used as an executable. Homebrew provides a `rewrite_shebang` method to rewrite the shebang of a script to point to the correct interpreter path. For example, the [`icdiff` formula](https://github.com/Homebrew/homebrew-core/blob/7beae5ab57c65249403699b2b0700fbccf14e6cb/Formula/icdiff.rb#L16) uses such utility.

Reviewer 2: It might be worth adding a mention of what it's rewritten from/to/why?

Author: I committed some changes to cover this.

The symlinks created by [`install_symlink`](https://rubydoc.brew.sh/Pathname#install_symlink-instance_method) are guaranteed to be relative. `ln_s` will only produce a relative symlink when given a relative path.

### Rewriting a script shebang

> Some formulae install executable scripts written in an interpreted language such as Python or Perl. Homebrew provides a `rewrite_shebang` method to rewrite the shebang of a script. This replaces a script's original interpreter path with the one installed by Homebrew and guarantees that the correct interpreter is used at execution time.

Reviewer 2: > This replaces a script's original interpreter path with the one installed by Homebrew

I'm not sure this is correct? It depends on the relevant OS, I think? CC @Bo98

Reviewer 3: Ah, yes. It's not entirely accurate. I think it depends on the dependencies a formula declares.

Reviewer 4: For Python it's right since we don't allow system Python. For Perl, it is indeed not correct as if `uses_from_macos "perl"` is used then it will use system Perl as the shebang on macOS.

It's perhaps also worth clarifying that `rewrite_shebang` is only needed where build system doesn't already do it (it's not needed under many `pip` and `ExtUtils::MakeMaker` cases).

### Rewriting a script shebang

> Some formulae install executable scripts written in an interpreted language such as Python or Perl. Homebrew provides a `rewrite_shebang` method to rewrite the shebang of a script. This replaces a script's original interpreter path with the one the formula depends on. This guarantees that the correct interpreter is used at execution time. It is only needed when the build system doesn't already do it (e.g., often when using `pip` or Perl `ExtUtils::MakeMaker`).

Reviewer: ```suggestion
Some formulae install executable scripts written in an interpreted language such as Python or Perl. Homebrew provides a `rewrite_shebang` method to rewrite the shebang of a script. This replaces a script's original interpreter path with the one the formula depends on. This guarantees that the correct interpreter is used at execution time. This isn't required if the build system already handles it (e.g. often with `pip` or Perl `ExtUtils::MakeMaker`).
```

Right now I think it's a little unclear whether `pip` and `ExtUtils::Maker` _do_ or _don't_ do it automatically.

> In some cases, the formula goal is to install a script written in an interpreted language, such as Python or Perl, to be used as an executable. Homebrew provides a `rewrite_shebang` method to rewrite the shebang of a script. This replaces the original interpreter path with the one installed by Homebrew. This guarantees that the interpreter dependency was correctly installed and is present at execution time. For example, the [`icdiff` formula](https://github.com/Homebrew/homebrew-core/blob/7beae5ab57c65249403699b2b0700fbccf14e6cb/Formula/icdiff.rb#L16) uses such utility. Note that it is necessary to include the utility in the formula, for example with Python one must use `include Language::Python::Shebang`.

Reviewer 3: ```suggestion
Some formulae install executable scripts written in an interpreted language such as Python or Perl. Homebrew provides a `rewrite_shebang` method to rewrite the shebang of a script. This replaces a script's original interpreter path with the one installed by Homebrew and guarantees that the correct interpreter is used at execution time.

For example, the [`icdiff` formula](https://github.com/Homebrew/homebrew-core/blob/7beae5ab57c65249403699b2b0700fbccf14e6cb/Formula/icdiff.rb#L16) uses such utility. Note that it is necessary to include the utility in the formula, for example with Python one must use `include Language::Python::Shebang`.
```
Let's shorten some of these sentences and split the paragraph up to help make it a bit more readable. (I won't mind if you prefer what you wrote originally, though.)

Author: I'm always happy to see any text I write improved :).
